tests/Doctrine/Tests/ORM/Functional/Ticket/GH9335Test.php

```diff
@@ -0,0 +1,220 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\DBAL\ParameterType;
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Doctrine\DBAL\Types\Type;
+use Doctrine\DBAL\Types\Type as DBALType;
+use Doctrine\ORM\Mapping\Column;
+use Doctrine\ORM\Mapping\Entity;
+use Doctrine\ORM\Mapping\Id;
+use Doctrine\ORM\Mapping\JoinColumn;
+use Doctrine\ORM\Mapping\OneToOne;
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+/**
+ * @group GH9335
+ */
+final class GH9335Test extends OrmFunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        if (! DBALType::hasType(GH9335IntObjectType::class)) {
+            DBALType::addType(GH9335IntObjectType::class, GH9335IntObjectType::class);
+        }
+
+        $this->setUpEntitySchema([GH9335Book::class, GH9335Author::class]);
+    }
+
+    /**
+     *  Verifies that entities with foreign keys with custom id object types don't throw an exception
+     *
+     * The test passes when refresh() does not throw an exception
+     */
+    public function testFlattenIdentifierWithObjectId(): void
+    {
+        $author = new GH9335Author('Douglas Adams');
+        $book   = new GH9335Book(new GH9335IntObject(42), 'The Hitchhiker\'s Guide to the Galaxy', $author);
+
+        $this->_em->persist($author);
+        $this->_em->persist($book);
+        $this->_em->flush($author);
+        $this->_em->flush($book);
+
+        $this->_em->refresh($book);
+
+        self::assertInstanceOf(GH9335IntObject::class, $book->getId());
+    }
+}
+
+
+class GH9335IntObjectType extends Type
+{
+    public function getSQLDeclaration(array $column, AbstractPlatform $platform): string
+    {
+        return $platform->getIntegerTypeDeclarationSQL($column);
+    }
+
+    public function getName(): string
+    {
+        return self::class;
+    }
+
+    public function convertToDatabaseValue($value, AbstractPlatform $platform): int
+    {
+        return $value->wrappedInt;
+    }
+
+    public function convertToPHPValue($value, AbstractPlatform $platform): GH9335IntObject
+    {
+        return new GH9335IntObject($value);
```

Comment:
I think you need to use a type cast here to get the tests back on track:

```suggestion
        return new GH9335IntObject((int) $value);
```

The DBAL does so, but I'm not sure why you get a string in the first place… I would have said it was because of the `STRINGIFY_FETCHES` [attribute](https://www.php.net/manual/en/pdo.setattribute.php), but since PHP 8.0, it defaults to `false`, so I don't understand, because the error also occurs on the PHP 8 builds.

Reply:
Thank you! 😊